File: server_refactor/api/organisms.py
from typing import Optional
from fastapi import APIRouter, Depends, Body
from services import organism_service

router = APIRouter()

class CommonQueryParams:
    def __init__(
        self,
        filter: Optional[str] = None,
        offset: int = 0,
        limit: int = 20,
        taxids: Optional[str] = None,
    ):
        self.filter = filter
        self.offset = offset
        self.limit = limit
        self.taxids = taxids

@router.get("/organisms")
@router.post("/organisms")
def get_organisms(commons: CommonQueryParams = Depends(), payload: Optional[dict] = Body(None)):
    """
    Get a list of organisms with optional filtering and pagination
    """
    if payload:
        params = payload
    else:
        params = commons.__dict__
    return organism_service.get_organisms(**params)

@router.get("/organisms/{taxid}")
def get_organism(taxid: str):
    """
    Get a single organism by taxid
    """
    return organism_service.get_organism(taxid).to_mongo().to_dict()

File: front/lib/api/organisms.ts
import { apiGet, type Query } from './base'
import type { OrganismRecord, Pagination } from './types'

export interface FetchOrganismsParams extends Query {
  filter?: string
  taxids?: string
  limit?: number
  offset?: number
}

export function listOrganisms(params: FetchOrganismsParams) {
  return apiGet<Pagination<OrganismRecord>>('/organisms', params)
}

export function getOrganism(taxid: string) {
  return apiGet<OrganismRecord>(`/organisms/${encodeURIComponent(taxid)}`)
}
